<a href="https://colab.research.google.com/github/Travistyse/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

Stretch goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Apply hypothesis testing to your personal project data (for the purposes of this notebook you can type a summary of the hypothesis you formed and tested)

In [0]:
### YOUR CODE STARTS HERE
import numpy as np
import pandas as pd
from scipy.stats import ttest_ind
from scipy.stats import ttest_1samp

pd.options.display.max_rows = 999
pd.options.display.max_columns = 100
# This example won't help me much as it's a 2 problem, not a single
single_example = "ttest_1samp(dem['bill'], .5, nan_policy='omit') where .5 is null hypothesis num"

Load and clean the data (or determine the best method to drop observations when running tests)

In [0]:
voting_url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data'
voting_data = pd.read_csv(voting_url)

In [4]:
voting_data.head()

republican  n  y n.1 y.1 y.2 y.3 n.2 n.3 n.4 y.4  ? y.5 y.6 y.7 n.5 y.8
0  republican  n  y   n   y   y   y   n   n   n   n  n   y   y   y   n   ?
1    democrat  ?  y   y   ?   y   y   n   n   n   n  y   n   y   y   n   n
2    democrat  n  y   y   n   ?   y   n   n   n   n  y   n   y   n   n   y
3    democrat  y  y   y   n   y   y   n   n   n   n  y   ?   y   y   y   y
4    democrat  n  y   y   n   y   y   n   n   n   n  n   n   y   y   y   y

In [8]:
len(voting_data)

434

In [0]:
rep = voting_data[voting_data['republican'] == 'republican']
dem = voting_data[voting_data['republican'] == 'democrat']

In [14]:
dem.head()

republican  n  y n.1 y.1 y.2 y.3 n.2 n.3 n.4 y.4  ? y.5 y.6 y.7 n.5 y.8
1   democrat  ?  y   y   ?   y   y   n   n   n   n  y   n   y   y   n   n
2   democrat  n  y   y   n   ?   y   n   n   n   n  y   n   y   n   n   y
3   democrat  y  y   y   n   y   y   n   n   n   n  y   ?   y   y   y   y
4   democrat  n  y   y   n   y   y   n   n   n   n  n   n   y   y   y   y
5   democrat  n  y   n   y   y   y   n   n   n   n  n   n   ?   y   y   y

In [16]:
# This matches what I was told - I've got the right data
# I'm now going to rename my data accordingly - not all of it
'''
1. Class Name: 2 (democrat, republican)
2. handicapped-infants: 2 (y,n)
3. water-project-cost-sharing: 2 (y,n)
4. adoption-of-the-budget-resolution: 2 (y,n)
5. physician-fee-freeze: 2 (y,n)
6. el-salvador-aid: 2 (y,n)
7. religious-groups-in-schools: 2 (y,n)
8. anti-satellite-test-ban: 2 (y,n)
9. aid-to-nicaraguan-contras: 2 (y,n)
10. mx-missile: 2 (y,n)
11. immigration: 2 (y,n)
12. synfuels-corporation-cutback: 2 (y,n)
13. education-spending: 2 (y,n)
14. superfund-right-to-sue: 2 (y,n)
15. crime: 2 (y,n)
16. duty-free-exports: 2 (y,n)
17. export-administration-act-south-africa: 2 (y,n)
'''
# If my goal was to simply find a divisive issue I would select 7.
# As such, and to challenge a bias of mine which states that repub. groups
# Would tend to be more in favor, I will select this one for my assignment.

dem_r = dem['y.3']
rep_r = rep['y.3']
print("r ", len(rep_r))
print("d ", len(dem_r))

r  167
d  267


In [17]:
# I currently - before removing null - have many more dem votes than r
len(rep_r) - rep_r.isnull().sum()

167

In [18]:
len(dem_r) - dem_r.isnull().sum()

267

This indicates that in my data there are no null values. Odd.

In [19]:
dem_r

1      y
2      y
3      y
4      y
5      y
8      n
11     n
12     y
15     y
16     n
18     n
19     ?
20     n
21     n
22     n
23     n
24     n
25     n
26     n
28     n
30     n
31     n
33     n
38     n
39     n
40     n
41     n
42     n
43     n
44     n
45     n
46     n
47     n
49     n
51     n
53     y
59     ?
61     n
62     n
63     n
67     n
68     n
69     n
71     n
73     y
74     y
75     y
76     y
77     y
79     y
80     n
84     y
87     y
89     n
90     n
91     n
92     n
93     y
94     y
95     y
96     y
97     y
99     y
100    y
101    y
102    y
103    y
104    n
107    n
108    n
109    n
111    y
113    n
114    ?
115    n
117    n
123    n
126    n
127    y
128    n
129    ?
130    y
136    y
137    y
138    y
142    n
143    y
144    y
146    n
148    n
150    y
151    y
152    y
156    y
158    y
159    y
160    y
161    y
163    y
164    y
167    y
168    n
169    n
171    n
172    y
173    n
174    n
176    n
177    n
178    n
179    n
1

That makes more sense. ? isn't null. Let's fix that using replace.

In [20]:
dem_r = dem_r.replace({'?':np.NaN, 'n':0, 'y':1})
rep_r = rep_r.replace({'?':np.NaN, 'n':0, 'y':1})
dem_r

1      1.0
2      1.0
3      1.0
4      1.0
5      1.0
8      0.0
11     0.0
12     1.0
15     1.0
16     0.0
18     0.0
19     NaN
20     0.0
21     0.0
22     0.0
23     0.0
24     0.0
25     0.0
26     0.0
28     0.0
30     0.0
31     0.0
33     0.0
38     0.0
39     0.0
40     0.0
41     0.0
42     0.0
43     0.0
44     0.0
45     0.0
46     0.0
47     0.0
49     0.0
51     0.0
53     1.0
59     NaN
61     0.0
62     0.0
63     0.0
67     0.0
68     0.0
69     0.0
71     0.0
73     1.0
74     1.0
75     1.0
76     1.0
77     1.0
79     1.0
80     0.0
84     1.0
87     1.0
89     0.0
90     0.0
91     0.0
92     0.0
93     1.0
94     1.0
95     1.0
96     1.0
97     1.0
99     1.0
100    1.0
101    1.0
102    1.0
103    1.0
104    0.0
107    0.0
108    0.0
109    0.0
111    1.0
113    0.0
114    NaN
115    0.0
117    0.0
123    0.0
126    0.0
127    1.0
128    0.0
129    NaN
130    1.0
136    1.0
137    1.0
138    1.0
142    0.0
143    1.0
144    1.0
146    0.0
148    0.0
150    1.0

In [21]:
# Now I can remove the NaNs properly
# Let's first make sure the vals changed
len(rep_r) - rep_r.isnull().sum()

165

In [22]:
len(dem_r) - dem_r.isnull().sum()

258

In [0]:
# Excellent, onto the removal
dem_r = dem_r[~np.isnan(dem_r)]

In [24]:
len(dem_r)

258

In [0]:
rep_r = rep_r[~np.isnan(rep_r)]

In [26]:
len(rep_r)

165

Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01

In [29]:
# Number 6 - el salvador aid - pretty positive dems support more than reps
# So I'll need to clean and gather that data
dem_es = dem['y.2']
rep_es = rep['y.2']
print("r null inc ", len(rep_es))
print("d null inc ", len(dem_es))

print('d_non_null ', len(dem_es) - dem_es.isnull().sum())
print('r_non_null ', len(rep_es) - rep_es.isnull().sum())

dem_es = dem_es.replace({'?':np.NaN, 'n':0, 'y':1})
rep_es = rep_es.replace({'?':np.NaN, 'n':0, 'y':1})

dem_es = dem_es[~np.isnan(dem_es)]
rep_es = rep_es[~np.isnan(rep_es)]

print("r null removed ", len(rep_es))
print("d null removed ", len(dem_es))

r null inc  167
d null inc  267
d_non_null  267
r_non_null  167
r null removed  164
d null removed  255


In [30]:
ttest_ind(rep_es, dem_es, nan_policy='omit')

Ttest_indResult(statistic=21.06444879561732, pvalue=1.3035583587181002e-67)

In [31]:
print( 'r ', rep_es.mean() )
print( 'd ', dem_es.mean() )

r  0.9512195121951219
d  0.21568627450980393


I pretty clearly see that, between the positive pvalue (which I'm having difficulty converting to a real number but atm assume is 1.verysmalldecimals ) and the significant difference in mean, against my initial thoughts, reps actually voted significantly more in favor of el salvador support than dems?

In [32]:
rep_es

0      1.0
6      1.0
7      1.0
9      1.0
10     1.0
13     1.0
14     1.0
17     1.0
27     1.0
29     1.0
32     1.0
34     1.0
35     1.0
36     1.0
37     1.0
48     1.0
50     1.0
52     1.0
54     1.0
55     1.0
56     1.0
57     1.0
58     1.0
60     1.0
64     1.0
65     1.0
66     1.0
70     0.0
72     1.0
78     1.0
81     1.0
82     1.0
83     1.0
85     1.0
86     1.0
88     1.0
98     1.0
105    1.0
110    1.0
112    1.0
116    1.0
118    1.0
119    1.0
120    1.0
121    1.0
122    1.0
124    1.0
125    1.0
131    1.0
132    1.0
133    1.0
134    1.0
135    1.0
139    0.0
140    1.0
141    1.0
145    1.0
147    1.0
149    1.0
153    1.0
154    1.0
155    1.0
157    1.0
162    1.0
165    1.0
166    1.0
170    1.0
175    0.0
187    1.0
189    1.0
190    1.0
194    1.0
196    1.0
203    1.0
205    1.0
206    1.0
210    1.0
213    1.0
216    1.0
222    1.0
223    1.0
224    1.0
227    1.0
228    1.0
229    1.0
230    1.0
232    1.0
234    1.0
238    1.0
239    0.0
241    1.0

In [33]:
dem_es

1      1.0
3      1.0
4      1.0
5      1.0
8      0.0
11     0.0
12     0.0
15     0.0
16     0.0
18     0.0
19     0.0
20     0.0
21     0.0
22     0.0
23     0.0
24     0.0
25     0.0
26     0.0
28     0.0
30     0.0
31     0.0
33     0.0
38     0.0
39     0.0
40     0.0
41     0.0
42     0.0
43     0.0
44     0.0
45     0.0
46     0.0
47     0.0
49     0.0
51     0.0
53     0.0
59     0.0
61     0.0
62     0.0
63     0.0
67     0.0
68     0.0
69     0.0
71     0.0
73     0.0
74     1.0
75     1.0
76     1.0
77     1.0
79     0.0
80     0.0
84     1.0
87     1.0
89     0.0
90     0.0
91     0.0
92     0.0
93     1.0
94     1.0
95     1.0
96     1.0
97     0.0
99     1.0
100    1.0
101    1.0
102    1.0
103    0.0
104    0.0
107    0.0
108    0.0
109    0.0
111    0.0
113    0.0
115    0.0
117    0.0
123    0.0
126    0.0
127    0.0
128    0.0
130    0.0
136    0.0
137    0.0
138    0.0
142    0.0
143    0.0
144    0.0
146    0.0
148    0.0
150    1.0
151    0.0
152    0.0
156    0.0

In [34]:
dem['y.2'] # Yeah, data lines up. 

1      y
2      ?
3      y
4      y
5      y
8      n
11     n
12     n
15     n
16     n
18     n
19     n
20     n
21     n
22     n
23     n
24     n
25     n
26     n
28     n
30     n
31     n
33     n
38     n
39     n
40     n
41     n
42     n
43     n
44     n
45     n
46     n
47     n
49     n
51     n
53     n
59     n
61     n
62     n
63     n
67     n
68     n
69     n
71     n
73     n
74     y
75     y
76     y
77     y
79     n
80     n
84     y
87     y
89     n
90     n
91     n
92     n
93     y
94     y
95     y
96     y
97     n
99     y
100    y
101    y
102    y
103    n
104    n
107    n
108    n
109    n
111    n
113    n
114    ?
115    n
117    n
123    n
126    n
127    n
128    n
129    ?
130    n
136    n
137    n
138    n
142    n
143    n
144    n
146    n
148    n
150    y
151    n
152    n
156    n
158    ?
159    y
160    y
161    y
163    y
164    n
167    y
168    n
169    n
171    n
172    y
173    n
174    n
176    n
177    n
178    n
179    n
1

Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01

In [27]:
# My hypothesis is that rep will support religious involvement in 
# Schools more than 50% and dem will support it less than 50%
# My Null hypothesis is, of course, that both support it equally - 
# or, perhaps, technically, the opposite would be possible but
# If anything that would be my alternative - not null, but a complete
# reversal of expectations.
ttest_ind(rep_r, dem_r, nan_policy='omit')

Ttest_indResult(statistic=9.694471882817886, pvalue=3.415425248480747e-20)

In [28]:
print( 'r ', rep_r.mean() )
print( 'd ', dem_r.mean() )

r  0.896969696969697
d  0.47674418604651164


In [0]:
# I clearly didn't need to test via Ttest as my hypothesis is clearly true
# Dems have a 47.674% average support which is less than 50% and less than:
# Reps (which) have a support avg of 89.6969% - significantly higher
# But with a p-value of 3.4154 I think this is very clearly in support
# of my alternative hypothesis - that reps support rel groups bill more


Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)
Note that this data will involve 2 sample t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

In [35]:
# Number 2 - infant dis
'''
Thought process:
Dems support because concern for citizens
Reps might support because babies but tbh bad and inconsistent logic has
been pointed out with the abortion rights movement so I wouldn't be
surprised if, in spite of how horrible it sounds on paper, reps don't
show support for disabled babies.

Giving them the benefit of the doubt in my hypothesis of equal support
'''
# So I'll need to clean and gather that data
dem_id = dem['n']
rep_id = rep['n']
print("r null inc ", len(rep_id))
print("d null inc ", len(dem_id))

print('d_non_null ', len(dem_id) - dem_id.isnull().sum())
print('r_non_null ', len(rep_id) - rep_id.isnull().sum())

dem_id = dem_id.replace({'?':np.NaN, 'n':0, 'y':1})
rep_id = rep_id.replace({'?':np.NaN, 'n':0, 'y':1})

dem_id = dem_id[~np.isnan(dem_id)]
rep_id = rep_id[~np.isnan(rep_id)]

print("r null removed ", len(rep_id))
print("d null removed ", len(dem_id))

r null inc  167
d null inc  267
d_non_null  267
r_non_null  167
r null removed  164
d null removed  258


In [36]:
ttest_ind(rep_id, dem_id, nan_policy='omit')

Ttest_indResult(statistic=-9.15392122841775, pvalue=2.4195550274147043e-18)

In [37]:
print( 'r ', rep_id.mean() )
print( 'd ', dem_id.mean() )

r  0.18902439024390244
d  0.6046511627906976


RIP. Kinda figured, but at least had some hope. The mean clearly shows that reps only have a supporting population of 18.9% whereas dems have 60.46%

The pvalue of 2.41955 is also pretty telling.

While I aimed to show an equal amount of support, it's pretty clear that the null hypothesis (that being equal support) has been rejected by this data.